EIPS/eip-4788.md

```diff
@@ -77,76 +79,163 @@ When verifying a block, execution clients **MUST** ensure the root value in the
+timestamp
+mod
+push3 0x018000
+add
+sstore
+stop
+```
 
-If the `timestamp` **does** match, the client **MUST** read the root from the contract storage and return those 32 bytes in the caller's return buffer.
+#### Deployment
+
+The beacon roots contract is deployed like any other smart contract. A special synthetic address is generated
+by working backwards from the desired deployment transaction:
+
+```json
+{
+  "type": "0x2",
+  "chainId": "0x1",
+  "nonce": "0x0",
+  "to": null,
+  "gas": "0xd4f8",
+  "gasPrice": null,
+  "maxPriorityFeePerGas": "0x9c7652400",
+  "maxFeePerGas": "0xe8d4a51000",
+  "value": "0x0",
+  "input": "0x605a8060095f395ff33373fffffffffffffffffffffffffffffffffffffffe14604257602036146024575f5ffd5b620180005f350680545f351415603d576201800001545f525b60205ff35b42620180004206555f3562018000420662018000015500",
+  "accessList": [],
+  "v": "0x0",
+  "r": "0x539",
+  "s": "0x1337",
+  "hash": "0x8ecfe5753922d27aa737597d946f638a15b7e3b5f74fef9ef8cf1b510a1af1cc"
+  }
+```
 
-In pseudocode:
+The sender of the transaction can be calculated as `0x01d0610058aC7AEF1887d8877ee7f04B7645Dc95`. The address of the first contract deployed from the account is `rlp([sender, 0])` which equals `0x502E02F5d91024A9AF0aB81fbF0a47Eb99a013aE`. Although this style of contract creation is not tied to any specific initcode like create2 is, the synthetic address is cryptographically bound to the input data of the transaction (e.g. the initcode).
 
-```python
-timestamp = evm.calldata[:32]
-if len(timestamp) != 32:
-    evm.revert()
-    return
+### Block processing
 
-timestamp_reduced = to_uint64_be(timestamp) % HISTORICAL_ROOTS_MODULUS
-timestamp_index = to_uint256_be(timestamp_reduced)
+At the start of processing any execution block where `block.timestamp >= FORK_TIMESTAMP` (i.e. before processing any transactions), call `BEACON_ROOTS_ADDRESS` as `SYSTEM_ADDRESS` with the 32-byte input of `header.parent_beacon_block_root`. This will trigger the `set()` routine of the beacon roots contract. This is a system operation and therefore:
 
-recorded_timestamp = sload(HISTORY_STORAGE_ADDRESS, timestamp_index)
-if recorded_timestamp != timestamp:
-    evm.returndata[:32].set(uint256(0))
-else:
-    timestamp_extended = timestamp_reduced + HISTORICAL_ROOTS_MODULUS
-    root_index = to_uint256_be(timestamp_extended)
-    root = sload(HISTORY_STORAGE_ADDRESS, root_index)
-    evm.returndata[:32].set(root)
-```
+* the call must execute to completion, therefore the available gas can be considered as infinite
+* the call does not count against the block's gas limit
+* the call does not follow the [EIP-1559](./eip-1559.md) burn semantics - no value should be transferred as part of the call
+* if no code exists at `BEACON_ROOTS_ADDRESS`, the call must fail silently
 
-The precompile costs `G_beacon_root` gas to reflect the two (2) implicit `SLOAD`s from the precompile's state.
+Client may decide to omit an explicit EVM call and directly set the storage values.
```

Comment:
This would make any potential modifications to the contract in the future harder or support some other contract on a different chain.

Reply:
Yes. This sentence should have some YMMV warnings: doing this may be a workable implementation of the mainnet contract, but might lead to issues on other networks, if some other contract is used. 

(The case where no contract at all exists should be fine, since setting the values directly doesn't affect the state -- the slots are cleaned up due to the `empty`-rule at the end of block anyway. )

Perhaps something like this? 

```suggestion
Clients may decide to omit an explicit EVM call and directly set the storage values. Note: While this is a valid optimization for Ethereum mainnet, it could be problematic on non-mainnet situations in case a different contract is used.
```